### PR TITLE
enhancement: add additional RF link info to the sysinfo.json file.

### PR DESCRIPTION
**Pull request by r1de**
Added: current TxMbps value (the negotiated modulation rate)
Added: current signal and noise levels
Added: current set frequency

### CODE REVIEW

**Review by ae6xe — CHANGES_REQUESTED** [review]
The bit rate returned is always for a 20MHz channel.  This value is not compensated for in iwinfo with different channel widths.    Need to divide by 2 if 10MHz and by 4 if 5MHz.    

However..  this value doesn't tell you the rate in use on a given link, it is probably only the highest rate in use somewhere.  If 10 links, there could be 10 different rates in use -- this is the transmit rate not receive.   If the purpose is to show rates on links, then you'd need to pull this from:

/sys/kernel/debug/ieee80211/phy0/netdev\:wlan0/stations/BSS addr of remote node/rc_stats

**Comment by r1de** [comment]
Ok. Thanks Joe. I did not know that it was showing some "maximum" bit rate (and I didn't even think of the bit rate for each link, duh!).  
How about we just add the Signal, Noise and Frequency and leave the TxMbps for later?  
There is talk of adding the LQ/NLQ info as well to sysinfo.json, that is also going to be a "per link" item, so it makes sense to have the bit rate there as well in the end.

**Comment by ae6xe** [comment]
A better option would be to pass back the actual throughput of data as show in this table (for each neighbor).   The 'A' row is the actual xmit thoughput (dividing by 2 or 4 for 10 and 5 MHz channels):

/sys/kernel/debug/ieee80211/phy0/netdev:wlan0/stations/MAC address of remote node/rc_stats

All the other measures (SNR, LQ, NLQ) are indirect and 'interesting' and also per neighbor.   If you translated the MAC to an IP address and sent back a table of "to_IP: throughput in mbps", and show on a map link would be the most meaningful.  This value makes different links directly comparable as to their performance.  A 'Quality' link is just one with a value above some throughput threshold.  The only other significant measure is latency of the link.   There's no place to pull this metric from that I know of -- has to be determined with a ping or similar method.

**Comment by r1de** [comment]
We can wait on all of it then.  
These additions to the sysinfo.json were requested by other users, they want to use them in some other scripts or something, I am not certain.  
I was also just trying to get the same values that the GUI had, just for consistency.  

Other than to just display this info and maybe use the frequency to check the band without searching through several channel lists, I really had no idea how I would use this on my map.  

But... you just gave me some other ideas, I think.  
I have to fully grok what you just said. :)  
Those rc_stats tables are quite interesting!  
I know I have seen that info before, but you just made it make way more sense! thanks!

**Comment by dman776** [comment]
Check the PR that adds iwinfo into the new /api page.   That may be a better place for this info.

Thanks,
Darryl

**Comment by r1de** [comment]
@ae6xe @dman776 Please test this, it's worked fine on the few nodes I have tried, but still.

**Comment by r1de** [comment]
Also Dependent on #389

**Comment by r1de** [comment]
I also moved the some of the sysinfo.json info into functions in the aredn/info.lua file.  
`services` and `hosts`  
This will hopefully help later.  

I can't seem to update that latest commit comment, please update if you can.

**Comment by r1de** [comment]
I see a couple of things I might have forgotten. Will update in just a bit.

File: files/usr/lib/lua/aredn/info.lua [hunk 1]
```diff
@@ -171,10 +173,199 @@ function model.getBand(radio)
...
+	freq=os.capture("iwinfo " .. wlanInf .. " info | egrep 'Mode:'")
+	freq=freq:gsub("^%s*(.-)%s*$", "%1")
+	freq=string.match(freq, "%((.-)%)")
+	return freq
+end
+
+-------------------------------------
+-- Return list of MAC to Hostname files
+-------------------------------------
+function mac2host(dir)
```
Comment (dman776) [inline files/usr/lib/lua/aredn/info.lua:190]: this would make more sense in the aredn.utils library

File: files/usr/lib/lua/aredn/info.lua [hunk 2]
```diff
@@ -171,10 +173,199 @@ function model.getBand(radio)
...
+		list[i] = filename
+	end
+	pfile:close()
+	return list
+end
+
+-------------------------------------
+-- Return Neighbor Link Info
+-------------------------------------
+function neighborLinkInfo()
```
Comment (dman776) [inline files/usr/lib/lua/aredn/info.lua:205]: this needs to be in the model scope.
ie.   function model.neighborLinkInfo()

File: files/usr/lib/lua/aredn/info.lua [hunk 3]
```diff
@@ -171,10 +173,199 @@ function model.getBand(radio)
...
+		neighborLinkInfo[name]["nlq"]=nlq
+		neighborLinkInfo[name]["link_type"]=linkType
+	end
+	return neighborLinkInfo
+end
+
+-------------------------------------
+-- Return locally hosted services (for sysinfo.json)
+-------------------------------------
+function local_services()
```
Comment (dman776) [inline files/usr/lib/lua/aredn/info.lua:261]: this needs to be in the model scope.

File: files/usr/lib/lua/aredn/info.lua [hunk 4]
```diff
@@ -171,10 +173,199 @@ function model.getBand(radio)
...
+		service['error']="Cannot read local services file"
+		table.insert(lclsrvs, service)
+	end
+	return lclsrvs
+end
+
+-------------------------------------
+-- Return *All* Network Services
+-------------------------------------
+function all_services()
```
Comment (dman776) [inline files/usr/lib/lua/aredn/info.lua:290]: this needs to be in the model scope.

File: files/usr/lib/lua/aredn/info.lua [hunk 5]
```diff
@@ -171,10 +173,199 @@ function model.getBand(radio)
...
+		service['error']="Cannot read services file"
+		table.insert(services,service)
+	end
+	return services
+end
+
+-------------------------------------
+-- Return *All* Hosts
+-------------------------------------
+function all_hosts()
```
Comment (dman776) [inline files/usr/lib/lua/aredn/info.lua:320]: this needs to be in the model scope.

File: files/usr/lib/lua/aredn/info.lua [hunk 6]
```diff
@@ -171,10 +173,199 @@ function model.getBand(radio)
...
+		host['error']="Cannot read hosts file"
+		table.insert(hosts,host)
+	end
+	info['hosts']=hosts
+end
+
+-------------------------------------
+-- Return link_info (for sysinfo.json)
+-------------------------------------
+function link_info()
```
Comment (dman776) [inline files/usr/lib/lua/aredn/info.lua:360]: this needs to be in the model scope.

File: files/www/cgi-bin/sysinfo.json [hunk 7]
```diff
@@ -147,95 +148,28 @@ info['interfaces']=tif
...
-				service['name']=name
-				table.insert(services,service)
-			end
-		end
-	else
-		service['error']="Cannot read services file"
-		table.insert(services,service)
-	end
-	info['services']=services
+	info['services']=all_services()
```
Comment (dman776) [inline files/www/cgi-bin/sysinfo.json:156]: aredn_info.all_services()

File: files/www/cgi-bin/sysinfo.json [hunk 8]
```diff
@@ -147,95 +148,28 @@ info['interfaces']=tif
...
-				service['link']=link
-				table.insert(lclsrvs, service)
-			end
-		end
-	else
-		service['error']="Cannot read local services file"
-		table.insert(lclsrvs, service)
-	end
-	info['services_local']=lclsrvs
+	info['services_local']=local_services()
```
Comment (dman776) [inline files/www/cgi-bin/sysinfo.json:161]: aredn_info.local_services()

File: files/www/cgi-bin/sysinfo.json [hunk 9]
```diff
@@ -147,95 +148,28 @@ info['interfaces']=tif
...
+end
+
+-- LOCAL NEIGHBORS LINK INFO
+if string.find(nixio.getenv("QUERY_STRING"):lower(),"link_info=1") then
+	info['link_info']=link_info()
+end
+
+-- MESHMAP INFO
+if string.find(nixio.getenv("QUERY_STRING"):lower(),"meshmap=1") then
+	info['services_local']=local_services()
```
Comment (dman776) [inline files/www/cgi-bin/sysinfo.json:171]: aredn_info.local_services()

File: files/www/cgi-bin/sysinfo.json [hunk 10]
```diff
@@ -147,95 +148,28 @@ info['interfaces']=tif
...
+	info['link_info']=link_info()
```
Comment (dman776) [inline files/www/cgi-bin/sysinfo.json:172]: aredn_info.link_info()

**Review by dman776 — CHANGES_REQUESTED** [review]
a few inline comments

**Comment by r1de** [comment]
Ok, I see all those and understand.  
Spent all morning trying to get a good firmware build with the lausocket stuff built in... git is a pain sometimes!  

I will resolve them once I get home.

**Comment by dman776** [comment]
Will resubmit with updated PR soon.

**Pull request closed without merging**